pkgs/top-level/all-packages.nix

```
@@ -2566,8 +2566,11 @@ with pkgs;
 
   microcom = callPackage ../applications/terminal-emulators/microcom { };
 
-  mlterm = darwin.apple_sdk_11_0.callPackage ../applications/terminal-emulators/mlterm {
-    inherit (darwin.apple_sdk_11_0.frameworks) Cocoa;
+  mlterm = callPackage ../applications/terminal-emulators/mlterm { };
+  mlterm-wayland = mlterm.override {
+    guiChoices = ["wayland"];
+    typeEnginesChoices = ["cairo"];
+    enableX = false;
```

W.r.t. no list args, how I'd like this to look like would be this:

```suggestion
  mlterm-wayland = mlterm.override {
    withWayland = true;
    withXorg = false;
```

Xft should automagically disable itself when !withXorg and withCairo should be on by default on any platform it's supported on.

---

> Xft should automagically disable itself when !withXorg and withCairo should be on by default on any platform it's supported on.

Could you rephrase that please?

---

Maybe he meant to say 
- withCairo should be true by default
  - or, worst hypothesis, false only in the case of not being supported
- `withXorg = false` should remove Xft from dependencies

---

> * `withXorg = false` should remove Xft from dependencies

I was thinking about this too, but for full user control, I don't want the expression to involve itself too much in logic that IMO should be upstream's responsibility. I think that if there are conflicting `./configure` flags, upstream should make it fail with an appropriate error message. This is also the same policy I practiced in GNURadio's expression, where there the logic of the different components' internal dependencies are much better documented, but much more complex. And indeed up to [some small mistakes](https://github.com/gnuradio/gnuradio/issues/3252), GNURadio is very responsible for this. With `mlterm` the relation between the different build features it's not well documented at all so I don't want to implement such logic based on Guess.

> `withCairo` should be true by default

`typeEnginesChoices.cairo = true;` by default (at least now).

---

> For a feature a user might want to wholly disable in all its forms (as in the case of X11), adding such a "master switch" makes sense. User turns off "X11 support" -> package turns off all X11-related features and removes xorg packages from inputs.

Using your `<xlib-feature> ? withX11` suggestion is indeed a good idea. Marking conversation as resolved.